src/unisparks/button.cpp

```diff
@@ -336,52 +341,46 @@ void doButtons(Player& player, const Milliseconds currentMillis) {
   uint8_t btn = buttonStatus(0, currentMillis);
 #if BUTTON_LOCK
   // info("doButtons start");
+  // 0 Locked and awaiting click; 1 Awaiting long press; 2 Awaiting click; 3 Awaiting long press; 4 Awaiting release; 5 Unlocked
   static uint8_t buttonLockState = 0;
-  static uint32_t lastButtonTime = 0;
+  static Milliseconds lockButtonTime = 0;
 
-  if (buttonLockState != 0 && currentMillis - lastButtonTime > lockDelay) {
+  // If idle-time expired, return to ‘locked’ state
+  if (buttonLockState != 0 && currentMillis - lockButtonTime >= 0) {
```

Comment:
```suggestion
  if (buttonLockState != 0 && currentMillis >= lockButtonTime) {
```

Reply:
We need to have a chat about modular arithmetic (and how to achieve that using the two’s complement signed arithmetic provided by today’s processors).

Testing “currentMillis >= lockButtonTime” fails when a signed integer wraps around — e.g. lockButtonTime is 0x7FFFFFFF (a big number) and currentMillis is two ticks later than that at 0x80000001 (a large negative number).

In that situation “currentMillis >= lockButtonTime” is false, which is not our intention.

Testing “currentMillis - lockButtonTime >= 0” gives the right answer: currentMillis is in fact two ticks later than the lockButtonTime.

Reply:
In this example, both `currentMillis` and `lockButtonTime` are of type `Milliseconds` which is typedefed to `int32_t`. Unlike unsigned integers, signed integers are not guaranteed to wrap by the C standard. As an example: the following code is undefined behavior:
```c
  Milliseconds m = 0x7FFFFFFF;
  m += 2;
  printf("%d\n", m);
```
which means that all bets are off if we let this code run for over 25 days. Maybe the ESP32 will use two's complement wrap around, maybe the compiler will do something else entirely. The most correct approach would be to trigger a restart if our implementation of `timeMillis()` ever reaches that state.

Reply:
I went ahead and implemented that in https://github.com/DavidSchinazi/unisparks/commit/f661f8b04d7cf6ced9d8c206a50868c3e12ee106